### PR TITLE
increased performance of k-diagonal extraction in da.diag() and da.diagonal()

### LINKED ISSUE #2726: Dask Array's diag supporting diagonal selection
Would be nice if Dask Array's [`diag`]( http://dask.pydata.org/en/latest/array-api.html#dask.array.diag ) supported the `k` value to allow selection of the diagonal as NumPy's implementation [does]( https://docs.scipy.org/doc/numpy-1.13.0/reference/generated/numpy.diag.html ). It seems like Dask Array's API docs suggest `diag` does support `k`, but the code does not allow for this option.

https://github.com/dask/dask/blob/f015b5f0808420bdd23f8408f5822d2f5aff8ff1/dask/array/creation.py#L413

### PR DESCRIPTION
**Pull request by ParticularMiner**
- [x] Closes #2726 (and maybe #5683)
- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

Perhaps you might find this useful.  It mirrors the `dask-grblas` implementation, which supports rectangular  chunks.  

It follows the straight path of the k-diagonal through the rectangular chunks of the input matrix , constructing the dask graph along the way.  So chunks untouched by the diagonal end up not being part of the final dask graph, reducing the algorithmic complexity of `diagonal(A, offset, axis1, axis2)` from 

O(*M*&nbsp;*N*<sub>axis1</sub>&nbsp;*N*<sub>axis2</sub>) &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;to &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; O(*M*&nbsp;max(1, *N*<sub>*k*</sub>)) ,

where *N*<sub>axis1</sub> (*N*<sub>axis2</sub>) is the number of chunks along `axis1` (`axis2`) of array `A`; *M* is the total number of chunks of `A` after `axis1` and `axis2` have been removed; while *N*<sub>*k*</sub> is the number of chunks touched by the k-diagonal after all axes, except `axis1` and `axis2`, have been removed. 

Relevant test-units have been modified to reflect this change. 

Critique is welcome.

NB: It might be worth comparing this with #5683 (which I only discovered after pushing this commit).

### CODE REVIEW

**Comment by GPUtester** [comment]
Can one of the admins verify this patch?

**Comment by quasiben** [comment]
add to allowlist

**Comment by TAdeJong** [comment]
At the time, #5683 failed on the issue described in #5661 with `test_corrcoef()` in `dask\array\tests\test_routines.py` (which was semi-unrelated). As of master today, #5661 still seems valid, so it might be interesting in comparing to see if this does not trigger #5661 and if so: why not.

**Comment by TAdeJong** [comment]
For completeness: 
I am the original author of #5683 and happy to let this PR supersede it :+1: 
As per discussion there, #5661 is essentially independent of this PR. (Which makes this all the better PR to merge of the 2).

File: dask/array/creation.py [hunk 1]
```diff
@@ -565,50 +565,114 @@ def eye(N, chunks="auto", M=None, k=0, dtype=float):
...
+            kdiag_row_stop = min(v.shape[0], v.shape[1] - k)
+            len_kdiag = kdiag_row_stop - kdiag_row_start
+            chunks = ((0,),) if len_kdiag <= 0 else ((len_kdiag,),)
+            dsk = {(name, 0): (np.diag, v, k)}
         else:
             raise ValueError("Array must be 1d or 2d only")
         return Array(dsk, name, chunks, meta=meta)
     if not isinstance(v, Array):
         raise TypeError(f"v must be a dask array or numpy array, got {type(v)}")
     if v.ndim != 1:
```
Comment (TAdeJong) [inline dask/array/creation.py:591]: Comparing to #5683, why don't you call da.diagonal here to reduce code duplication?

(And maybe combine #8703 into this)

**Comment by TAdeJong** [comment]
Other than thinking about not duplicating code from `da.diagonal` as mentioned above, this PR seems in all ways better than #5683 to me.
I think (admittedly from the sidelines, it has been some time since I was this deep in the dask code base) that if @ParticularMiner already had a good reason to not call `da.diagonal` this can be merged, #5683 can be closed/rejected and #8703 can be handled separately, otherwise it might be worth it to incorporate #8703 here, call `da.diagonal` and review and merge this as one PR.

**Comment by ParticularMiner** [comment]
Btw, many thanks for your constructive critique and support of this PR @TAdeJong .

File: dask/array/creation.py (hunk 1 above)
Comment (ParticularMiner) [reply dask/array/creation.py:590]: Done!

File: dask/array/tests/test_creation.py [hunk 2]
```diff
@@ -407,32 +407,34 @@ def test_eye():
...
 
     x = np.arange(64).reshape((8, 8))
-    assert_eq(da.diag(x), np.diag(x))
-
+    assert_eq(da.diag(x, k), np.diag(x, k))
     d = da.from_array(x, chunks=(4, 4))
-    assert_eq(da.diag(d), np.diag(x))
+    assert_eq(da.diag(d, k), np.diag(x, k))
+    d = da.from_array(x, chunks=((3, 2, 3), (4, 4)))
+    assert_eq(da.diag(d, k), np.diag(x, k))
```
Comment (jcrist) [inline dask/array/tests/test_creation.py:437]: Can you add a test with a rectangular input array as well? I only see square input arrays above.

While you're at it, would you mind splitting up the tests here a bit to separate the two uses of `diag`? Putting them together in the same test makes it a bit hard to read.

- One (or more, up to you) tests for creating a diagonal array from a 1d array
- One (or more, up to you) tests for extracting the diagonal from a 2d array

**Review by jcrist — COMMENTED** [review]
Thanks @ParticularMiner , overall this looks good to me. I left one test comment, but the rest looks good. While you're looking at this code, can I ask that you look through the tests for `da.diagonal` as well to see if they offer full coverage for the different parameters we'd care about here (dimensions of input array, chunking variation, axis arguments, etc...)? It'd be good to ensure we have full coverage here while you're still thinking about this code.

File: dask/array/tests/test_creation.py (hunk 2 above)
Comment (ParticularMiner) [reply dask/array/tests/test_creation.py:437]: Good point.  I'll do that.

Comment (ParticularMiner) [reply dask/array/tests/test_creation.py:464]: I've added and reorganized the tests plus ensured full coverage of `diag()` & `diagonal()`

Comment (ParticularMiner) [reply dask/array/tests/test_creation.py:463]: @jcrist 

CI tests failures seem to reveal that `cupy` type input arrays to `diagonal()` need special handling.  I'm taking a closer look at this.

Comment (ParticularMiner) [reply dask/array/tests/test_creation.py:463]: @jcrist

CI tests for `cupy`-type arrays have now passed!

**Review by jcrist — APPROVED** [review]
:rocket:

**Comment by jakirkham** [comment]
Thanks all! 😄

**Pull request merged**